### PR TITLE
feat(ngFor): add custom trackBy function support

### LINKED ISSUE #4402: [ngFor] trackby equivalent ?
I don't think there is an equivalent of AngularJS 1.x `trackBy` for `ngFor` right now (alpha-37). 
Is there a way to do the same (maybe using a custom `IterableDiffers`, but that looks like a bit difficult...), or is it a feature planned for later?

### PR DESCRIPTION
**Pull request by kara**
Make it possible to track items in iterables in custom ways (e.g. by ID or index), rather than simply by identity.

Usage:

With template element:

```
@Component(
  template: `
   <template ngFor #item [ngForOf]="items" [ngForTrackBy]=customTrackBy">
      {{item}}
   </template>
   `
)
class MyComponent {
  customTrackBy(index: number, obj: any): any {
    return index;
  }
}
```

With *ngFor:

```
@Component(
  template: `
   <div *ngFor="#item of items; trackBy:customTrackBy">
      {{item}}
   </div>
   `
)
class MyComponent {
  customTrackBy(index: number, obj: any): any {
    return index;
  }
}
```

@mhevery  Did you want to review this?

Closes #4402

### CODE REVIEW

File: modules/angular2/src/core/change_detection/differs/default_iterable_differ.ts [hunk 1]
```diff
@@ -42,6 +43,10 @@ export class DefaultIterableDiffer implements IterableDiffer {
   private _removalsHead: CollectionChangeRecord = null;
   private _removalsTail: CollectionChangeRecord = null;
 
+  constructor(private _trackByFn?: TrackByFn) {
```
Comment (vsavkin) [inline modules/angular2/src/core/change_detection/differs/default_iterable_differ.ts]: Can you do `_trackByFn: TrackByFn = trackByIdentity` instead? Try to do this if ts2dart supports it. If not, what you have is good.

Comment (kara) [inline modules/angular2/src/core/change_detection/differs/default_iterable_differ.ts]: Yeah, that's what I wanted to do too :)  It doesn't seem to be supported.

File: modules/angular2/test/core/change_detection/differs/default_iterable_differ_spec.ts [hunk 2]
```diff
@@ -324,5 +324,100 @@ export function main() {
...
+            .toEqual(iterableChangesAsString({
+              collection:
+                  ['{\n  "id": "b"\n}[1->0]', '{\n  "id": "a"\n}[0->1]', '{\n  "id": "c"\n}'],
+              previous: ['{\n  "id": "a"\n}[0->1]', '{\n  "id": "b"\n}[1->0]', '{\n  "id": "c"\n}'],
+              moves: ['{\n  "id": "b"\n}[1->0]', '{\n  "id": "a"\n}[0->1]']
+            }));
+
+      });
+
+      it('should track duplicate reinsertion normally with track by function', () => {
```
Comment (vsavkin) [inline modules/angular2/test/core/change_detection/differs/default_iterable_differ_spec.ts]: I am confused about this case. Could you explain why we need to support this?

File: modules/angular2/test/core/change_detection/differs/default_iterable_differ_spec.ts [hunk 3]
```diff
@@ -324,5 +324,100 @@ export function main() {
...
     });
+
+    describe('trackBy function', function() {
+      var differ;
+      var trackByItemId =
+          (index: number, item: any): any => { return StringMapWrapper.get(item, 'id') }
+
+      var createItem =
+          (val: string) => {
+            var item = StringMapWrapper.create();
```
Comment (vsavkin) [inline modules/angular2/test/core/change_detection/differs/default_iterable_differ_spec.ts]: You don't have to use StringMapWrapper. You could just use

```
var item = {'id': val};
``

Comment (kara) [inline modules/angular2/test/core/change_detection/differs/default_iterable_differ_spec.ts]: I was getting a weird Dart error, so I ended up switching to StringMapWrapper.  I think the problem might have been that I was omitting the quotation marks around ID. Changing!

File: modules/angular2/test/core/change_detection/differs/default_iterable_differ_spec.ts (hunk 2 above)
Comment (kara) [inline modules/angular2/test/core/change_detection/differs/default_iterable_differ_spec.ts]: This is something we were already supporting for regular track-by-identity (with the `verifyReinsertion` code) to ensure that duplicates don't get re-ordered in unexpected ways.  I thought it should probably work the same way even when tracking by ID, so added a test.

**Comment by johnpapa** [comment]
I see this unit test, but I cannot seem to get ngForTrackBy to work for my sample.

https://github.com/angular/angular/blob/cee2318110eeea115e5f6fc5bfc814cbaa7d90d8/modules/angular2/test/common/directives/ng_for_spec.ts#L364-L402

I added `[ngForTrackBy]="customTrackBy"` to my element in the template. Then I added the `customTrackBy` function the way the unit test shows it. But I am getting this error:

```
EXCEPTION: Template parse errors:
Can't bind to 'ngForTrackBy' since it isn't a known native property ("    <div class="mdl-cell mdl-cell--3-col"
        *ngFor="#character of characters | async"
        [ERROR ->][ngForTrackBy]="trackBy"
        (click)="gotoDetail(character)">

"): DashboardComponent@6:8
```

It seems it wants me to declare `ngForTrackBy` to the component, but I do not see where it is exported from any module by Angular. Nor is the unit test I referenced doing that.

What else is needed?

cc // @kara @wardbell

**Pull request closed without merging**